src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -107,7 +107,10 @@ object RuleUtils {
       indexes.filter(index =>
         index.created && isHybridScanCandidate(index, filesByRelations.flatten))
     } else {
-      indexes.filter(index => index.created && signatureValid(index))
+      indexes.filter(
+        index =>
+          index.created && signatureValid(index) &&
+            index.deletedFiles.isEmpty && index.appendedFiles.isEmpty)
```

Comment:
This would be ok but might cause some bad user experience:

1) refreshIncremental
2) failed to apply index
3) no idea or clue for the failure until the user checks index log entry ..

Reply:
Thanks @sezruby for bringing this up, i think that's ok. This behavior is expected. I think this experience is exactly same as the following:
1. user creates index
2. user updates source data
3. fail to apply index
4. no idea or clue for the failure until the user checks index log entry ..

Please feel free to suggest how to improve this experience. We can create an issue for the same and fix it in subsequent PRs.

On the other hand, if we don't do this, it will lead to incorrect results.

Reply:
The user experience currently is not good either. For example, with hybrid scan off, if you add a file, it will fail in `signatureValid` without letting the user know. We have been talking about exposing "why not" API that tells the user why an index was not picked up. Until we introduce that kind of API, I think the user experience will remain "not desirable" - meaning not good. 😄

Reply:
I think @apoorvedave1 beat me on this comment 😄

Reply:
@imback82 @apoorvedave1 
This is why I suggested quick failure (assert) rather than storing deletedFiles but maybe we could hybrid scan with it later. 

Now we are proposing "mutable" dataset, I think we need to address this issue with more care. 

Anyway I'm fine with the current approach as it's also valid :) 

Thanks all!

Reply:
This check will be needed for quick mode (metadata only update) anyway, no?

Reply:
Another way to avoid the check is to change the semantics of the Source.Content.

Currently, the source files and their signature is based on the "latest" snapshot when the action was run. "appendedFiles" and "deletedFiles" are the new/deleted files "since" the index "data" was created.

We moved to this semantics while reviewing #170, where signature was based on the "latest" but source files were "since".

Another way is to make everything "since" such that:
The sources files and their signature is "when" the index "data" was created. "appendedFiles" and "deletedFiles" are the new/deleted files "since" the index "data" was created.

Can we discuss pros/cons for each approach?

(For example, are we losing anything if we move the current approach (mix of "latest" and "since") to the new approach (everything "since")? I think the current approach is to favor hybrid scan such that if the signature matches, we can just utilize "appendedFiles" and "deletedFiles" - which is not yet implemented btw.)

Reply:
In my perspective, I prefer everything "since" for the incremental mode - and not allow deleted or appended files in the metadata like full mode. But for this we need to merge RefreshAppend and RefreshDelete. 

For quick mode, we could use the current approach - update signature with the latest and collect appended and deleted files. 

This is what I was thinking first and I think we could avoid possible complex cases and conditions with this approach.

Reply:
Can we first agree on what we want to store in the metadata? How we handle the metadata in different refresh modes is an orthogonal discussion. (For example, we should store the same info (everything "since" or mix) regardless of refresh modes).

Reply:
> This check will be needed for quick mode (metadata only update) anyway, no?

for continuing with this discussion, I don't think we need this check for metadtata only update either. Could you please give me a scenario?
I was thinking metadata only update is much simpler. with `RefreshMetadataAction` (or something I read in another PR), we can just update the 'deleted' and 'appended' files in metadata. Isn't that enough? Why do we need the assert(deleted.nonEmpty) I don't really understand.

Reply:
I meant "This check" = ` index.deletedFiles.isEmpty && index.appendedFiles.isEmpty)` in this file, not the assert. What I mean is that with the current semantics, we need this check regardless of the refresh modes (which are orthogonal).

Reply:
yeah these checks are required with the current semantics regardless of the refresh mode. So here's my understanding:

If hybrid scan is enabled, we know the data-source-changes (either adds or deletes) will be handled by hybrid scan.
If it is not, the right way to handle this is DO NOT USE the index because it's not up to date with the data.

So I am currently fine with this semantic.

Reply:
Sorry I misread. I thought "since" means "latest". So "latest"-`source.Content` indicates the source files of the index data in that version. Then I think "mixed" approach is better as there is no special use of "since" "appendedFiles" and "deletedFiles" during query time or refresh action?

Reply:
Oh OK. The use of "latest" could be confusing. Just to make sure everyone is on the same page, let's assume the following events in sequence:
- There are source files: f1, f2, f3
- An index is created.
- A source file is added: f4
- A source file is deleted: f1
- Refresh "quick" (metadata only operation)

With the current semantics (mix of "latest" and "since"), we will have the following in the latest index log entry:
- source files / signature: f2, f3, f4 / sig(f2, f3, f4)
- appendedFiles: f4 
- deletedFiles: f1

With the semantics of always using "since":
- source files / signature: f1, f2, f3 / sig(f1, f2, f3)
- appendedFiles: f4 
- deletedFiles: f1

And I am hearing that the current semantics (mix of "latest" and "since") will be a better use. Is that correct?

Reply:
That's correct. The benefit of `quick` is skipping the comparison of file list for hybrid scan and using the signature instead. ( + and it also allows selective hybrid scan - disable hybrid scan config and do quick refresh only for the indexes which are candidates of hybrid scan op. )
On the other hand, with always using "since" approach, we need to check if the source file list in the given plan could be the candidate of each index in some how at query execution time.

Reply:
What if we capture the latest source signature as separate field? For example, we introduce `Update` case class that captures "appendedFiles", "deletedFiles", and "latestSourceFilesSignature"' for the `Source.Content`? (I am just exploring the best option for both semantics / usage before the next release, so please bear with me 😄 ). It seems that if the signature doesn't match, we need to fallback to iterating files for hybrid scan anyway.


> ( + and it also allows selective hybrid scan - disable hybrid scan config and do quick refresh only for the indexes which are candidates of hybrid scan op. )

Can you also elaborate on this scenario? I didn't get the part where you disable the hybrid scan config and do quick refresh since the paths are different (one for query and another for non-query).

Reply:
> What if we capture the latest source signature as separate field? For example, we introduce Update case class that captures "appendedFiles", "deletedFiles", and "latestSourceFilesSignature"' for the Source.Content? 

This could be an option though the original signature value won't be used anymore.
But in this way, we could keep the same semantic for Source.Content. 
And in the case of "full" and "incremental" refresh, `Source.Content` contains the source files at refresh (both full and incremental) the same as now?

> Can you also elaborate on this scenario? I didn't get the part where you disable the hybrid scan config and do quick refresh since the paths are different (one for query and another for non-query).

For example, there're 5 indexes but one of them has many "diff" files in its source files since the index build. But a user want to use hybrid scan for an index with very little diff files. If global hybrid scan config is enabled, Hyperspace tries to apply hybrid scan for all candidate indexes - if there's no proper hybrid scan threshold in candidate selection. 
So `quick` refresh can be an option for this situation:
- disable hybrid scan config
- quick refresh for some indexes with little diff
- hybrid scan is applied only for `quick` refreshed indexes

Although this is different context, I just wanted to describe the possible use case of `quick` refresh.

Reply:
@imback82 @sezruby Should we move this discussion to an issue so we can follow-up properly through a PR, if necessary? If you agree, @sezruby - can you please create an issue with all the approaches + pros/cons so everyone is on the same page? You can paste the link to that issue here as a final reply.

Reply:
> This could be an option though the original signature value won't be used anymore.

Yes, it will be used only if `Update` is `isEmpty == true` (assuming we can capture `Update` as `Option[]`).

> But in this way, we could keep the same semantic for Source.Content.

Yes, I think the code logic will be simpler as well.

> And in the case of "full" and "incremental" refresh, `Source.Content` contains the source files at refresh (both full and incremental) the same as now?

Well, it will store the source files that are indexed, so it could be different if there are deleted files before `RefreshAppend` runs.

> Although this is different context, I just wanted to describe the possible use case of `quick` refresh.

Oh, you mean refresh "on-the-fly" during query time.

Reply:
Created #198 to track this issue. 
BTW
>>Although this is different context, I just wanted to describe the possible use case of quick refresh.
>
>Oh, you mean refresh "on-the-fly" during query time.

it's possible but I meant a user can choose the candidate indexes for hybrid scan - with disabled global config.

Reply:
OK, in this case, I still don't understand 😄. @rapoth do you understand this scenario?

Reply:
Ok let me try again 😁 
1. disable "**global**" hybrid scan config
2. user can choose some "beneficial - with high perf improvement" indexes with little diff and do `quick` refresh for the indexes
3.  hybrid scan can be performed for `quick` refreshed indexes even if the global config is disabled. The other unrefreshed indexes with diff won't be the candidates for hybrid scan.

Reply:
Thanks 👍 . So in a nutshell, the selective hybrid scan is "we can apply the hybrid scan even if the user didn't enable it only when the signature matches", right?

Reply:
Right. That's the point 👍

Reply:
Thanks!